src/DeltaLake/Kernel/Arrow/Handlers/ArrowFFIInterOpHandler.cs

```diff
@@ -188,64 +157,46 @@ private static unsafe RecordBatch AddPartitionColumnsToRecordBatch(RecordBatch r
                         ptr = (sbyte*)partitionCols->Cols[i],
                         len = (ulong)colName?.Length
                     },
                     Marshal.GetFunctionPointerForDelegate<AllocateStringFn>(StringAllocatorCallbacks.AllocateString)
                 );
-                string partitionVal = partitionValPtr != null ? Marshal.PtrToStringAnsi((IntPtr)partitionValPtr) : String.Empty;
+                string colVal = colValPtr != null ? Marshal.PtrToStringAnsi((IntPtr)colValPtr) : String.Empty;
 #pragma warning restore CS1024, CS8629, CS8600
 
-                for (int j = 0; j < recordBatch.Length; j++)
+                if (!string.IsNullOrEmpty(colName) && !string.IsNullOrEmpty(colVal))
+                {
+                    colNames.Add(colName);
+                    colValues.Add(colVal);
+                }
+                else
                 {
-                    columnBuilder = columnBuilder.Append(partitionVal ?? "");
+                    throw new InvalidOperationException($"Failed to parse partition columns (got {colName}) and values (got {colVal}) from the Kernel");
                 }
-                columns.Add(columnBuilder.Build());
             }
-            return new RecordBatch(schemaBuilder.Build(), columns, recordBatch.Length);
-        }
 
-#pragma warning disable CA1859, IDE0060 // Although we're not using partitionValues right now, it will be used when Kernel supports reporting Arrow Schema
-        private static unsafe IArrowType DeterminePartitionColumnType(string colName, CStringMap* partitionValues)
-        {
-            // Currently, there's no way to determine the type of the partition,
-            // because the Kernel always represents partition values as strings in CStringMap.
-            //
-            // We have a request with Kernel team here to get back the Arrow Schema from
-            // the Delta Transaction Log:
-            //
-            // >>> https://delta-users.slack.com/archives/C04TRPG3LHZ/p1728001059452499?thread_ts=1727999835.930339&cid=C04TRPG3LHZ
-            //
-            return StringType.Default;
-        }
-#pragma warning restore CA1859, IDE0060
+            char** colNamesPtr = (char**)Marshal.AllocHGlobal(colNames.Count * sizeof(char*));
+            char** colValuesPtr = (char**)Marshal.AllocHGlobal(colValues.Count * sizeof(char*));
 
-        private static unsafe Apache.Arrow.Schema AddPartitionColumnsToSchema(Apache.Arrow.Schema originalSchema, PartitionList* partitionCols, CStringMap* partitionValues)
-        {
-            Apache.Arrow.Schema.Builder schemaBuilder = new();
-            foreach (Field field in originalSchema.FieldsList)
+            for (int i = 0; i < colNames.Count; i++)
             {
-                schemaBuilder = schemaBuilder.Field(field);
+                colNamesPtr[i] = (char*)Marshal.StringToHGlobalAnsi(colNames[i]);
+                colValuesPtr[i] = (char*)Marshal.StringToHGlobalAnsi(colValues[i]);

```

Comment:
Again, should probably be UTF8 and not Ansi. Unfortunately, there's no StringToHGlobalUTF8 (though there is a StringToCoTaskMemUTF8, which is what we've used for ADBC). I think under recent version of Windows there may not be a difference, but it's easy enough to implement a StringToHGlobalUTF8 by hand.

Reply:
So this one is interesting, it looks like they hid the UTF8 method under `StringToHGlobalAuto`:

https://github.com/dotnet/runtime/blob/43a9236abfd8e44154b54cc61afdc13e81932aa9/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Unix.cs#L21-L24

![image](https://github.com/user-attachments/assets/eb1c1f23-e441-46ce-a2be-c0f2deb62f7c)

Discussion:

https://github.com/dotnet/runtime/issues/99886

The interesting part is, when I change the code to use that:

![image](https://github.com/user-attachments/assets/480ad25d-6c34-418e-bda5-32b116fcb1fc)

The test fails on the DataFrame conversion:

![image](https://github.com/user-attachments/assets/da704c75-119b-415f-97e1-60ea294316b6)

But if I revert the code back to `StringToHGlobalAnsi`, test runs fine (both Windows and Linux):

Windows:
![image](https://github.com/user-attachments/assets/deded68a-8dca-4684-9fde-3b6740519195)

Linux:
![image](https://github.com/user-attachments/assets/c2c2a5d0-cf78-498d-a8cc-0a6436855c26)

I understand you're definitely speaking from experience on localization - I have very little experience :)
Any thoughts on what could be the issue with `StringToHGlobalAnsi` (and what would an alternative be - if not `StringToHGlobalAuto`)?

Reply:
First off, ideally any testing you do that involves marshaling strings should include code points > 127 as that's where the behavior between different encodings starts to diverge. Any such code points will do; I typically resort to a handful of German characters such as `ßäöü` because I know how to get them on a US keyboard so I can do it quickly.

I believe `StringToHGlobalAuto` has OS-dependent behavior -- ANSI on Windows and UTF-8 on Unix. But we always want UTF-8 no matter what the OS because both Rust and Delta use UTF-8 strings on all OSes. (And .NET strings are of course UTF-16 on all OSes.)

I have the vague understanding that on modern Windows there's no difference between HGLOBAL-allocated memory and CoTaskMemAlloc'd memory, so it ought to be okay to use StringToCoTaskMemUTF8 for this conversion. In fact, in the ADBC code we have a downlevel implementation of this function which I just noticed yesterday uses AllocHGlobal instead of CoTaskMemAlloc and this has not yet caused a problem. But if you want to ensure that the memory was allocated from the "HGLOBAL" heap and not the "CoTaskMemAlloc" heap you can see how it's done by looking at the implementation of `StringToCoTaskMemUTF8` in [https://github.com/apache/arrow-adbc/blob/main/csharp/src/Apache.Arrow.Adbc/Extensions/MarshalExtensions.cs](https://github.com/apache/arrow-adbc/blob/main/csharp/src/Apache.Arrow.Adbc/Extensions/MarshalExtensions.cs).

Reply:
Thank you @CurtHagenlocher! 

`StringToCoTaskMemUTF8` worked great, tested on both Windows and Linux. The CI should also test in Mac.

Windows:

![image](https://github.com/user-attachments/assets/3b4192b6-2cc1-4e1e-a9fb-49ee68de573c)

Linux:

![image](https://github.com/user-attachments/assets/218a8d27-cd16-4166-bc34-2464553e863b)

I'll be sure to come to you on any localization problems in future 😉